### ISSUE TITLE
Stack level too deep

### ISSUE THREAD

**Issue by mizujin** [Shopify/theme-check#165]
- ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-darwin18]
- theme-check 0.3.2

Hello! Using theme-check via CLI or Vscode, getting an error on some project folders, not all of them.

```
stack level too deep
/usr/local/Cellar/ruby/3.0.0_1/lib/ruby/3.0.0/set.rb:130:in `initialize_dup'
/usr/local/Cellar/ruby/3.0.0_1/lib/ruby/3.0.0/set.rb:461:in `dup'
/usr/local/Cellar/ruby/3.0.0_1/lib/ruby/3.0.0/set.rb:461:in `|'
/usr/local/Cellar/theme-check/0.3.2/gems/theme-check-0.3.2/lib/theme_check/checks/unused_assign.rb:14:in `block in collect_used_assigns'
/usr/local/Cellar/ruby/3.0.0_1/lib/ruby/3.0.0/set.rb:344:in `each_key'
/usr/local/Cellar/ruby/3.0.0_1/lib/ruby/3.0.0/set.rb:344:in `each'
/usr/local/Cellar/theme-check/0.3.2/gems/theme-check-0.3.2/lib/theme_check/checks/unused_assign.rb:12:in `collect_used_assigns'
/usr/local/Cellar/theme-check/0.3.2/gems/theme-check-0.3.2/lib/theme_check/checks/unused_assign.rb:14:in `block in collect_used_assigns'
```

My local project config file only contains an ignore rule for node_modules/*

**Comment by macournoyer** [comment]
Interesting! I think there might be a recursion in one of your snippets (you `{% include ... %}` a snippet that ends up `includ`ing itself).

But it should definitely not crash because of this. I'll fix this.

And we should also add a new check to detect recursion!

**Comment by mizujin** [comment]
After checking, I can confirm that this is recursion happening in a snippet that includes itself in a loop. This is the base code from a Shopify paid theme, Story.

In snippets/nav-item.liquid : 

```
    <div class="main-menu-dropdown" id="{{ link_id }}">
      <ul>
        {% for link in link.links %}
          {% include 'nav-item' %}
        {% endfor %}
      </ul>
    </div>
``` 

If this is bad practice, it should definitely raise a warning! Thanks for checking this out.

**Comment by macournoyer** [comment]
Oh my bad! The recursion does make sense when rendering the theme. It's definitely an issue in theme check only.

**Comment by charlespwd** [comment]
Fix was released in v0.3.3. 

```
brew upgrade theme-check
```
